packages/core/src/subscriptions/index.ts

```diff
@@ -0,0 +1,259 @@
+ *
+ * -----
+ *
+ * ### Events emitted:
+ *
+ * - `connect` - A new subscription is connected to the `SubscriptionManager` and `message` events for this subscription can be expected.
+ * - `disconnect` - The specified subscription is no longer being monitored by the `SubscriptionManager`.
+ * - `error` - An error has occurred.
+ * - `message` - A message containing a notification `Bundle` has been received.
+ * - `close` - The WebSocket has been closed.
+ * - `heartbeat` - A `heartbeat` message has been received.
+ */
+export class SubscriptionEmitter extends TypedEventTarget<SubscriptionEventMap> {
+  private criteria: Set<string>;
+  constructor(...criteria: string[]) {
+    super();
+    this.criteria = new Set(criteria);
+  }
+  getCriteria(): Set<string> {
+    return this.criteria;
+  }
+  [kAddCriteria](criteria: string): void {
+    this.criteria.add(criteria);
+  }
+  [kRemoveCriteria](criteria: string): void {
+    this.criteria.delete(criteria);
+  }
+}
+
+export class SubscriptionManager {
+  private readonly medplum: MedplumClient;
+  private ws: WebSocket;
+  private masterSubEmitter?: SubscriptionEmitter;
+  private subEmitters: Map<string, SubscriptionEmitter>; // Map<criteria, SubscriptionEmitter>
+  private refCounts: Map<string, number>; // Map<criteria, refCount>
+  private subscriptionCriteriaLookup: Map<string, string>; // Map<subscriptionId, criteria>
+  private criteriaSubscriptionLookup: Map<string, string>; // Map<criteria, subscriptionId>
+  private wsClosed: boolean;
+
+  constructor(medplum: MedplumClient, wsOrUrl: WebSocket | string) {
+    if (!(medplum instanceof MedplumClient)) {
+      throw new OperationOutcomeError(validationError('First arg of constructor should be a `MedplumClient`'));
+    }
+    let ws: WebSocket;
+    if (typeof wsOrUrl === 'string') {
+      let url: string;
+      try {
+        url = new URL(wsOrUrl).toString();
+      } catch (_err) {
+        throw new OperationOutcomeError(validationError('Not a valid URL'));
+      }
+      ws = new WebSocket(url);
+    } else if (!(wsOrUrl instanceof WebSocket)) {
+      throw new OperationOutcomeError(validationError('Invalid WebSocket'));
+    } else {
+      ws = wsOrUrl;
+    }
+
+    this.medplum = medplum;
+    this.ws = ws;
+    this.masterSubEmitter = new SubscriptionEmitter();
+    this.subEmitters = new Map<string, SubscriptionEmitter>();
+    this.refCounts = new Map<string, number>();
+    this.subscriptionCriteriaLookup = new Map<string, string>();
+    this.criteriaSubscriptionLookup = new Map<string, string>();
+    this.wsClosed = false;
+
+    this.setupWebSocketListeners();
+  }
+
+  private setupWebSocketListeners(): void {
+    const ws = this.ws;
+
+    ws.addEventListener('message', (event: MessageEvent) => {
+      const bundle = JSON.parse(event.data) as Bundle;
+      // Get criteria for event
+      const status = bundle?.entry?.[0]?.resource as SubscriptionStatus;
```

Comment:
1. We should probably treat this as `SubscriptionStatus | undefined` to be safe.  Maybe we can count on Medplum server to always include the status, but probably not other servers.
2. Out of curiosity, what happens if this callback function throws?  Does it gracefully recover, or does it kill the connection?

Reply:
1. It's part of the [spec](https://build.fhir.org/bundle.html#subscription-notification:~:text=A%20subscription%20notification%20(type%20%3D%20%22subscription%2Dnotification%22)%20consists%20of%20a%20series%20of%201%20or%20more%20entries%2C%20the%20first%20of%20which%20is%20a%20SubscriptionStatus.) that `SubscriptionStatus` is the first element in the bundle, and without the `SubscriptionStatus` it's not possible to resolve which callback should be called. We can handle it as an error but it complicates the code and would only be possible if communicating with a non-compliant server.

Reply:
2. This is typically handled by the `EventEmitter` (or `EventTarget`) implementation, but that's a valid question. I'll add a test for that

Reply:
Apparently `EventTarget` lets uncaught exceptions get thrown to `UncaughtException` handler, so I don't think we should add logic like that (Node's `EventEmitter` does emit errors to `error` event listeners if they exist before throwing them to the `UncaughtException` handler).

Instead I wrapped everything in a try catch within the handler and emitted `error` from our callback in the case of something like an invalid message being received from the WebSocket.